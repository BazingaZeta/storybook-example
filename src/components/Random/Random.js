import React from 'react'
import PropTypes from 'prop-types'

import Button from 'components/Button'

import { Wrapper, H1 } from '../typography'

const Random = ({
  isLoading,
  hasError,
  isFulfilled,
  number,
  onClick,
  header,
  label,
  actionLabel,
  loadingLabel
}) => {
  /** Define pristine state condition, when user didn't do any actions */
  const isPristine = !isLoading && !hasError && !isFulfilled

  return (
    <Wrapper>
      <H1>{header}</H1>
      <Button disabled={isLoading} onClick={onClick} primary size="large" label={label} />
      {isPristine && <div>{actionLabel}</div>}
      {isLoading && <div>{loadingLabel}</div>}
      {isFulfilled && (
        <div>
          Retrieved: <tt>{number}</tt>
        </div>
      )}
      {hasError && <div>Error: Could not retrieve a random number</div>}
    </Wrapper>
  )
}

Random.propTypes = {
  isLoading: PropTypes.bool,
  hasError: PropTypes.bool,
  isFulfilled: PropTypes.bool,
  number: PropTypes.number,
  onClick: PropTypes.func.isRequired,
  header: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  actionLabel: PropTypes.string.isRequired,
  loadingLabel: PropTypes.string.isRequired
}

Random.defaultProps = {
  isLoading: false,
  hasError: false,
  isFulfilled: false,
  number: null
}

export default Random
