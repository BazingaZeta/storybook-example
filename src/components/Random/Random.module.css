.counter {
  border: 1px solid lightgray;
  margin: 36px 24px;
  padding: 24px;
  text-align: center;
  width: 240px;
}

.header {
  font-size: 24px;
  font-weight: normal;
  margin: 0 0 12px;
}
