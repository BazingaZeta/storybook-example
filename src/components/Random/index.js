import React from 'react'
import { useActions, useRandomAPI } from '../../features/random'
import PureRandom from './Random'

const Random = () => {
  /**
   *  Get number, returned  from random.org, and the state of request from Redux store.
   */
  const { number, isLoading, hasError, isFulfilled } = useRandomAPI()

  /** Create incrementCounter action, using custom hook from feature */
  const { getNumber } = useActions()

  return (
    <PureRandom
      isLoading={isLoading}
      hasError={hasError}
      isFulfilled={isFulfilled}
      header="Random Number"
      number={number}
      actionLabel="Click the button to get random number"
      loadingLabel="Getting number"
      onClick={getNumber}
      label="Get random number"
    />
  )
}

export default Random
