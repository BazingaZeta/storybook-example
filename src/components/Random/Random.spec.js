import React from 'react'
import { render, fireEvent } from '@testing-library/react'

import Random from './Random'

describe('components > Random > Pure', () => {
  const onClick = jest.fn()
  const header = 'some header'
  const label = 'some label'
  const number = 555
  const actionLabel = 'some actionLabel'
  const loadingLabel = 'some loadingLabel'

  /**
   * Provide table of values to run tests with
   * @see https://jestjs.io/docs/en/api#describeeachtablename-fn-timeout
   */
  describe.each`
    isLoading | hasError | isFulfilled
    ${false}  | ${false} | ${false}
    ${true}   | ${false} | ${false}
    ${false}  | ${true}  | ${false}
    ${false}  | ${false} | ${true}
  `('renders different store states', ({ isLoading, hasError, isFulfilled }) => {
    it(`when isLoading=${isLoading} && hasError=${hasError} && isFulfilled=${isFulfilled}`, () => {
      const { asFragment } = render(
        <Random
          isLoading={isLoading}
          hasError={hasError}
          isFulfilled={isFulfilled}
          header={header}
          number={number}
          actionLabel={actionLabel}
          loadingLabel={loadingLabel}
          onClick={onClick}
          label={label}
        />
      )

      /**
       * Basic snapshot test to check, if rendered component
       * matches expected footprint.
       */
      expect(asFragment()).toMatchSnapshot()
    })
  })

  describe('on button click', () => {
    let component
    let button

    beforeAll(() => {
      component = render(
        <Random
          isFulfilled
          header={header}
          number={number}
          actionLabel={actionLabel}
          loadingLabel={loadingLabel}
          onClick={onClick}
          label={label}
        />
      )

      button = component.getByRole('button')
      fireEvent.click(button)
    })

    it('invoked the onClick handler', () => {
      expect(onClick).toHaveBeenCalledTimes(1)
    })
  })
})
