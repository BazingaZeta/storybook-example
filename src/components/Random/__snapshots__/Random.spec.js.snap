// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components > Random > Pure renders different store states when isLoading=false && hasError=false && isFulfilled=false 1`] = `
<DocumentFragment>
  <div
    class="sc-gtssRu evpqoY"
  >
    <h1
      class="sc-dlnjPT bgFMnT"
    >
      some header
    </h1>
    <button
      class="sc-bdnylx lolljv"
      type="button"
    >
      some label
    </button>
    <div>
      some actionLabel
    </div>
  </div>
</DocumentFragment>
`;

exports[`components > Random > Pure renders different store states when isLoading=false && hasError=false && isFulfilled=true 1`] = `
<DocumentFragment>
  <div
    class="sc-gtssRu evpqoY"
  >
    <h1
      class="sc-dlnjPT bgFMnT"
    >
      some header
    </h1>
    <button
      class="sc-bdnylx lolljv"
      type="button"
    >
      some label
    </button>
    <div>
      Retrieved: 
      <tt>
        555
      </tt>
    </div>
  </div>
</DocumentFragment>
`;

exports[`components > Random > Pure renders different store states when isLoading=false && hasError=true && isFulfilled=false 1`] = `
<DocumentFragment>
  <div
    class="sc-gtssRu evpqoY"
  >
    <h1
      class="sc-dlnjPT bgFMnT"
    >
      some header
    </h1>
    <button
      class="sc-bdnylx lolljv"
      type="button"
    >
      some label
    </button>
    <div>
      Error: Could not retrieve a random number
    </div>
  </div>
</DocumentFragment>
`;

exports[`components > Random > Pure renders different store states when isLoading=true && hasError=false && isFulfilled=false 1`] = `
<DocumentFragment>
  <div
    class="sc-gtssRu evpqoY"
  >
    <h1
      class="sc-dlnjPT bgFMnT"
    >
      some header
    </h1>
    <button
      class="sc-bdnylx lolljv"
      disabled=""
      type="button"
    >
      some label
    </button>
    <div>
      some loadingLabel
    </div>
  </div>
</DocumentFragment>
`;
